src/ngamsServer/ngamsServer/ngamsServer.py

```diff
@@ -436,6 +435,68 @@ def proxy_request(self, host_id, host, port, timeout=300):
 
         self.send_data(data, mime_type, code=code, hdrs=hdrs)
 
+    def remote_proxy_request(self, request, host, port, timeout=300):
+        """Proxy the current request to remote host ``host``:``port``"""
```

Comment:
This method looks very similar to the "GET" portion of `proxy_request`, but maybe they are sufficiently different? In particular in `proxy_request` we don't handle the internal response as a stream but accumulate it all in memory instead, so it's much nicely done here. I still wonder whether there is some value in trying to merge these two.

Reply:
You are correct. There is some overlap between `proxy_request()` and `remote_proxy_request()`. The `proxy_request()` does a few things differently. For example,  like waking up a sleeping host in the cluster. This does not apply to the `remote_proxy_request()`.

I was a little surprised when when I looked at the `remote_proxy()`. Is it safe to store the contents of a whole file in memory? On ALMA manage some very large files in order of GBs. I would expect this could cause NGAS to crash when memory allocation is exceeded. It also looks better when you are measuring download progress using CURL. It avoids a big delay at the start because the file is being buffered in memory before it is streamed. It is therefore more efficient.

In ALMA I believe we configure NGAS to redirect instead of proxying requests in the local cluster. I should probably setup a test for this and see what happens.

I think the `write_data()` method would benefit from streaming chunks of data. We could refactor this in a separate step.

Reply:
> I was a little surprised when when I looked at the remote_proxy()

Your surprise is shared. This code was written like that a long time ago, and while I might have moved it around I never got around to behave properly (stream instead of accumulate). This is definitely a problem, but probably the reason why we haven't tackled it is because our installations don't use proxying, hence we are avoiding the problem. I'll open an issue to work on this ASAP.

Given all my comments in https://github.com/ICRAR/ngas/pull/28#discussion_r490725897, it looks more and more to me that `remote_proxy_request` is a specific case of `proxy_request`, if the latter was correctly implemented using streaming, and had the options to opt out from a few things. Once I implement these fixes I can go and do this little refactoring though, there's no need to hold this PR back (and having tests will definitely help to ensure I'm not breaking things!).

Reply:
OK, I've created #29 to track the issue with `proxy_request`.